### DISCUSSION TITLE
Add a way for a client to trigger unloading an index without restarting the server altogether

### DISCUSSION THREAD

**Discussion by tristone13th** [clangd/clangd#1454] [Q&A]
workspace/symbol should return the symbols for the current workspace, but when I switch to a new c workspace, I can also find the symbols of the original workspace in the result.

**Comment by HighCommander4** [comment]
What editor do you use? How are you switching to a new workspace?

If the "switch workspace" operation does not involve restarting the clangd process, that could explain the observed behaviour.

**Reply by tristone13th** [reply]
Hi, I use neovim and switch workspace by simpling changing the pwd, I suppose that would be enough because clangd can detect the workspace by root patterns (e.g., .git, compile_commands.json), if each time I switch workspace by restarting the clangd process, that would be too heavy. Can you offer me some suggestions? thanks!

**Reply by HighCommander4** [reply]
As far as I'm aware, the current working directory is a per-process piece of state, and there is no way for clangd, being a child process of the editor, to know that the editor's current working directory has changed. I think the editor would have to notify clangd about this by sending a notification over the Language Server Protocol, and I don't think there's currently an LSP message for this. (It's possible I'm overlooking something.)

In addition, clangd deliberately supports keeping multiple indexes loaded at the same time, since indexes are stored next to the compile_commands.json, and it's somewhat common for larger projects to contain multiple subprojects or dependent projects with separate compile_commands.json files (and thus separate indexes), where it's useful to cross-reference symbols between the different indexes.

I suppose it would be possible to add a protocol extension for unloading the index of a particular directory, as a lighter-weight alternative to restarting clangd altogether. We'd have to figure out how this would fit into different editors' notions of projects/workspaces, i.e. when it would make sense for editors to perform this, keeping in mind we'd also like to keep the multiple-subprojects use case working. (And then of course editor plugins would have to be modified to use this protocol extension.)

My best suggestion in the meantime is to restart clangd when switching workspaces; in my experience, this is usually pretty quick. Alternatively, if you have a small number of workspaces you switch between frequently, you could consider having a different editor instance (each using its own clangd process) open for each.

**Reply by tom-anders** [reply]
@tristone13th you could probably do this automatically by using the [DirChanged](https://neovim.io/doc/user/autocmd.html#Dirchanged) autocmd to trigger :LspRestart

**Reply by tristone13th** [reply]
> As far as I'm aware, the current working directory is a per-process piece of state, and there is no way for clangd, being a child process of the editor, to know that the editor's current working directory has changed. I think the editor would have to notify clangd about this by sending a notification over the Language Server Protocol, and I don't think there's currently an LSP message for this. (It's possible I'm overlooking something.)
> 
> In addition, clangd deliberately supports keeping multiple indexes loaded at the same time, since indexes are stored next to the compile_commands.json, and it's somewhat common for larger projects to contain multiple subprojects or dependent projects with separate compile_commands.json files (and thus separate indexes), where it's useful to cross-reference symbols between the different indexes.
> 
> I suppose it would be possible to add a protocol extension for unloading the index of a particular directory, as a lighter-weight alternative to restarting clangd altogether. We'd have to figure out how this would fit into different editors' notions of projects/workspaces, i.e. when it would make sense for editors to perform this, keeping in mind we'd also like to keep the multiple-subprojects use case working. (And then of course editor plugins would have to be modified to use this protocol extension.)
> 
> My best suggestion in the meantime is to restart clangd when switching workspaces; in my experience, this is usually pretty quick. Alternatively, if you have a small number of workspaces you switch between frequently, you could consider having a different editor instance (each using its own clangd process) open for each.

Thanks, I think adding a protocol extension would be help because restarting clangd each time is a bit slower, for example, sometimes I will switch between Linux kernel and qemu, these 2 are both very large projects and it will take some time to load the LSP entirely. In my experience, about 10 seconds.

**Reply by tristone13th** [reply]
> @tristone13th you could probably do this automatically by using the [DirChanged](https://neovim.io/doc/user/autocmd.html#Dirchanged) autocmd to trigger :LspRestart

Thanks, this will help but restarting LSP each time when switching dir is not suitable for very large projects such as Linux kernel and qemu.
